services/cache/lru_striped.go

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package cache
+
+import (
+	"math"
+	"runtime"
+	"time"
+
+	"github.com/cespare/xxhash/v2"
+)
+
+// LRUStriped keeps LRU caches in buckets in order to lower mutex contention.
+// This is achieved by hashing the input key to map it to a dedicated bucket.
+// Each bucket (an LRU cache) has it's own lock that helps distributing the lock
+// contention on multiple threads/cores, leading to less wait times.
+//
+// LRUStriped implements the Cache interface with the same behavior as LRU.
+//
+// Note that, because of it's distributed nature, the fixed size cannot be strictly respected
+// and you may have a tiny bit more space for keys than you defined through LRUOptions.
+// Bucket size is computed as follows: (size / nbuckets) + (size % nbuckets)
+//
+// Because of this size limit per bucket, and because of the nature of the data, you
+// may have buckets filled unevenly, and because of this, keys evicted from the entire cache
+// where an simple LRU wouldn't have. Example
+//
+// Two buckets B1 and B2, of max size 2 each, meaning, theoretically, a max size of 4:
+//  * Say you have a set of 3 keys, they could fill an entire LRU cache.
+//  * But if all those keys are assigned to a single bucket B1, the first key will be evicted from B1
+//  * B2 will remain empty, even though there was enough memory allocated
+//
+// With 4 buckets and random UUIDs as keys, the amount of false evictions is around 5%.
+//
+// By default, the number of buckets equals the number of cpus returned from runtime.NumCPU.
+//
+// This struct is lock-free and intended to be used without lock.
+type LRUStriped struct {
+	buckets []*LRU
```

Comment:
There might be less pointer chasing going on, if we have `[]LRU` instead and return a pointer to the entire `LRUStriped` struct. It should be more cache friendly as a whole.

Reply:
i remember doing that and iirc there was no improvement 🤔 i will give it another shot

Reply:
ah no i remember now: it’s pointless to try doing so because then we will have the lock as pointer, otherwise we are copying the lock and that makes any vet/lint tool cry hard, which is legit

Reply:
Here is the diff i’m using:

```
diff --git a/services/cache/lru.go b/services/cache/lru.go
index a580296b3..7e802d88a 100644
--- a/services/cache/lru.go
+++ b/services/cache/lru.go
@@ -53,14 +53,17 @@ type entry struct {
 
 // NewLRU creates an LRU of the given size.
 func NewLRU(opts LRUOptions) Cache {
-	return &LRU{
-		name:                   opts.Name,
-		size:                   opts.Size,
-		evictList:              list.New(),
-		items:                  make(map[string]*list.Element, opts.Size),
-		defaultExpiry:          opts.DefaultExpiry,
-		invalidateClusterEvent: opts.InvalidateClusterEvent,
-	}
+	return (&LRU{}).init(opts)
+}
+
+func (l *LRU) init(opts LRUOptions) Cache {
+	l.name = opts.Name
+	l.size = opts.Size
+	l.evictList = list.New()
+	l.items = make(map[string]*list.Element, opts.Size)
+	l.defaultExpiry = opts.DefaultExpiry
+	l.invalidateClusterEvent = opts.InvalidateClusterEvent
+	return l
 }
 
 // Purge is used to completely clear the cache.
diff --git a/services/cache/lru_striped.go b/services/cache/lru_striped.go
index 02feb9229..6275d47be 100644
--- a/services/cache/lru_striped.go
+++ b/services/cache/lru_striped.go
@@ -37,7 +37,7 @@ import (
 //
 // This struct is lock-free and intended to be used without lock.
 type LRUStriped struct {
-	buckets []*LRU
+	buckets []LRU
 }
 
 func (L LRUStriped) hashkeyMapHash(key string) uint64 {
@@ -45,15 +45,15 @@ func (L LRUStriped) hashkeyMapHash(key string) uint64 {
 }
 
 func (L LRUStriped) keyBucket(key string) *LRU {
-	return L.buckets[L.hashkeyMapHash(key)%uint64(len(L.buckets))]
+	return &L.buckets[L.hashkeyMapHash(key)%uint64(len(L.buckets))]
 }
 
 // Purge loops through each LRU cache for purging. Since LRUStriped doesn't use any lock,
 // each LRU bucket is purged after another one, which means that keys could still
 // be present after a call to Purge.
 func (L LRUStriped) Purge() error {
-	for _, lru := range L.buckets {
-		lru.Purge() // errors from purging LRU can be ignored as they always return nil
+	for i := 0; i < len(L.buckets); i++ {
+		_ = (&L.buckets[i]).Purge() // errors from purging LRU can be ignored as they always return nil
 	}
 	return nil
 }
@@ -88,8 +88,8 @@ func (L LRUStriped) Remove(key string) error {
 // This is not as precise as using a single LRU instance.
 func (L LRUStriped) Keys() ([]string, error) {
 	keys := make([]string, 0)
-	for _, lru := range L.buckets {
-		k, _ := lru.Keys() // Keys never returns any error
+	for i := 0; i < len(L.buckets); i++ {
+		k, _ := (&L.buckets[i]).Keys() // Keys never returns any error
 		keys = append(keys, k...)
 	}
 	return keys, nil
@@ -98,8 +98,8 @@ func (L LRUStriped) Keys() ([]string, error) {
 // Len does the same as LRU.Len. As for LRUStriped.Keys, this call cannot be precise.
 func (L LRUStriped) Len() (int, error) {
 	size := 0
-	for _, lru := range L.buckets {
-		s, _ := lru.Len() // Len never returns any error
+	for i := 0; i < len(L.buckets); i++ {
+		s, _ := (&L.buckets[i]).Len() // Len never returns any error
 		size += s
 	}
 	return size, nil
@@ -138,9 +138,10 @@ func NewLRUStriped(opts LRUOptions) Cache {
 	// now this is the size for each bucket
 	opts.Size = (opts.Size / opts.StripedBuckets) + (opts.Size % opts.StripedBuckets)
 
-	buckets := make([]*LRU, 0)
+	buckets := make([]LRU, 0, opts.StripedBuckets)
 	for i := 0; i < opts.StripedBuckets; i++ {
-		buckets = append(buckets, NewLRU(opts).(*LRU))
+		buckets = append(buckets, LRU{})
+		(&buckets[i]).init(opts)
 	}
 
 	return LRUStriped{
diff --git a/services/cache/lru_striped_bench_test.go b/services/cache/lru_striped_bench_test.go
index cd184964e..de9a411eb 100644
--- a/services/cache/lru_striped_bench_test.go
+++ b/services/cache/lru_striped_bench_test.go
@@ -14,7 +14,7 @@ import (
 )
 
 const (
-	M              = 500_000
+	M              = 10_000_000
 	BucketAssigned = 0
 )
 ```

i increased the number of iterations in the benchmark for both []LRU and []*LRU and it gives the same results

Reply:
We don't need to change to pointer everywhere. We can just change the receiver to use a pointer. Let me check that.

Reply:
LRU receiver methods are already using a pointer ^^

am i missing something?

Reply:
Duh sorry about that. Obviously that won't work. This is fine. :+1: